EEJT-0003/src/main/java/com/epam/esm/giftcertficate/service/GiftCertificateService.java

```diff
@@ -0,0 +1,220 @@
+        }
+        log.info("Transaction has been ended");
+        log.error("Rollback");
+        log.error("This gift certificate with (name = " + giftCertificate.getName() + ") has already existed");
+        throw new ServerException("this gift certificate with (name = " + giftCertificate.getName() + ") has already existed");
+    }
+
+    public Page<GiftCertificate> getAll(int page, int size) {
+        log.info("Service receives params for getting all");
+
+        PageRequest pageRequest = PageRequest.of(page, size);
+
+        log.debug("Service returns all gift-certificates");
+        return giftCertificateRepository.findAll(pageRequest);
+    }
+
+    public GiftCertificate getOneGiftCertificateById(long id) {
+        log.info("Service receives params for getting gift-certificate by id");
+        log.debug("Service returns one gift-certificate");
+
+        return giftCertificateRepository.findById(id).orElseThrow(
+                () -> {
+                    log.error("there are no gift certificate with (id = " + id + ")");
+                    return new ItemNotFoundException("there are no gift certificate with (id = " + id + ")");
+                });
+    }
+
+    @Transactional
+    @Modifying
+    public GiftCertificate updateGiftCertificate(long id, Set<Tag> tags, Map<String, String> updatesMap) {
+        log.info("Service receives gift-certificate for updating");
+        log.info("Transaction has been started");
+
+        updatesMap.remove(TAGS);
+
+        log.debug("Checking that gift-certificate has not already exist to update on new name");
+        if (!giftCertificateRepository.isGiftCertificateExistByName(updatesMap.get(NAME))) {
+
+            log.debug("Update");
+
+            GiftCertificate updatedGiftCertificate = getOneGiftCertificateById(id);
+
+            setFieldsForUpdatingNotNull(updatesMap.get("name"), updatedGiftCertificate::setName);
+            setFieldsForUpdatingNotNull(updatesMap.get("description"), updatedGiftCertificate::setDescription);
+
+            if (updatesMap.get("price") != null) {
+                setFieldsForUpdatingNotNull(Integer.parseInt(updatesMap.get("price")), updatedGiftCertificate::setPrice);
+            }
+            if (updatesMap.get("duration") != null) {
+                setFieldsForUpdatingNotNull(Integer.parseInt(updatesMap.get("duration")), updatedGiftCertificate::setDurationDays);
+            }
+
+            log.debug("Checking tags of gift-certificate");
+            if (tags != null) {
+
+                log.debug("Updating tags of gift-certificate");
+                if (!updatedGiftCertificate.getTags().isEmpty()) {
+                    updatedGiftCertificate.setTags(null);
+                }
+                if (!tags.isEmpty()) {
+                    updatedGiftCertificate.setTags(tags);
+                    tagService.verifyIsTagsExistWhenCreatingOrUpdatingGiftCertificate(updatedGiftCertificate.getTags());
+                }
+            }
```

Comment:
IMO for the sake or readability the "tag" part can be moved into separate method

Reply:
done